Factory/HttpFoundationFactory.php

```diff
@@ -0,0 +1,199 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function createRequest(ServerRequestInterface $psrRequest)
+    {
+        $parsedBody = $psrRequest->getParsedBody();
+        $parsedBody = is_array($parsedBody) ? $parsedBody : array();
+
+        $request = new Request(
+            $psrRequest->getQueryParams(),
+            $parsedBody,
+            $psrRequest->getAttributes(),
+            $psrRequest->getCookieParams(),
+            $this->getFiles($psrRequest->getUploadedFiles()),
+            $psrRequest->getServerParams(),
+            $psrRequest->getBody()->__toString()
+        );
+        $request->headers->replace($psrRequest->getHeaders());
+
+        return $request;
+    }
+
+    /**
+     * Converts to the input array to $_FILES structure.
+     *
+     * @param array $uploadedFiles
+     *
+     * @return array
+     */
+    private function getFiles(array $uploadedFiles)
+    {
+        $files = array();
+
+        foreach ($uploadedFiles as $key => $value) {
+            if ($value instanceof UploadedFileInterface) {
+                $files[$key] = $this->createUploadedFile($value);
+            } else {
+                $files[$key] = $this->getFiles($value);
+            }
+        }
+
+        return $files;
+    }
+
+    /**
+     * Creates Symfony UploadedFile instance from PSR-7 ones.
+     *
+     * @param UploadedFileInterface $psrUploadedFile
+     *
+     * @return UploadedFile
+     */
+    private function createUploadedFile(UploadedFileInterface $psrUploadedFile)
+    {
+        $temporaryPath = $this->getTemporaryPath();
+        $psrUploadedFile->moveTo($temporaryPath);
+
+        $clientFileName = $psrUploadedFile->getClientFilename();
+
+        return new UploadedFile(
+            $temporaryPath,
+            null === $clientFileName ? '' : $clientFileName,
+            $psrUploadedFile->getClientMediaType(),
+            $psrUploadedFile->getSize(),
+            $psrUploadedFile->getError(),
+            true
+        );
+    }
+
+    /**
+     * Gets a temporary file path.
+     *
+     * @return string
+     */
+    protected function getTemporaryPath()
+    {
+        return tempnam(sys_get_temp_dir(), uniqid('symfony', true));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createResponse(ResponseInterface $psrResponse)
+    {
+        $response = new Response(
+            $psrResponse->getBody()->__toString(),
+            $psrResponse->getStatusCode(),
+            $psrResponse->getHeaders()
+        );
+        $response->setProtocolVersion($psrResponse->getProtocolVersion());
+
+        foreach ($psrResponse->getHeader('Set-Cookie') as $cookie) {
+            $response->headers->setCookie($this->createCookie($cookie));
+        }
+
+        return $response;
+    }
+
+    /**
+     * Creates a Cookie instance from a cookie string.
+     *
+     * Some snippets have been taken from the Guzzle project: https://github.com/guzzle/guzzle/blob/5.3/src/Cookie/SetCookie.php#L34
```

Comment:
technically, we already have a Set-Cookie parsing logic in the BrowserKit component as it deals with the client side